diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index d79118c..0f6ffcb 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -42,7 +42,10 @@
 #include "cpudeadline.h"
 #include "cpuacct.h"
 #include "features.h"
-
+static inline struct cfs_rq *cfs_rq_of(struct sched_entity *se)
+{
+	return se->cfs_rq;
+}
 #ifdef CONFIG_SCHED_DEBUG
 # define SCHED_WARN_ON(x)	WARN_ONCE(x, #x)
 #else
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index f2568ef..f886371 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -344,10 +344,7 @@ static inline struct cfs_rq *task_cfs_rq(struct task_struct *p)
 }
 
 /* runqueue on which this entity is (to be) queued */
-static inline struct cfs_rq *cfs_rq_of(struct sched_entity *se)
-{
-	return se->cfs_rq;
-}
+
 
 /* runqueue "owned" by this group */
 static inline struct cfs_rq *group_cfs_rq(struct sched_entity *grp)
